Isolate matching code

sgossage opened the pull request:
- [ ] Isolate code relevant to matching
- [ ] Test everything works

mkruckow commented:
The matching should not go into `girds`, it belongs to `interpolation`.

sgossage commented:
I will update this PR with Camille's changes in [PR#521](https://github.com/POSYDON-code/POSYDON/pull/521)

mkruckow commented on posydon/grids/track_match.py:
```diff
@@ -829,8 +829,20 @@ def get_MESA_labels(list_for_matching):
+                
+                percent_diff = [
+                                "% difference",
+                                f'{100.0*abs(star.mass - self.get_track_val("mass", htrack, *best_sol.x))/star.mass :.3f}', 
+                                f'{100.0*abs(star.log_R - self.get_track_val("log_R", htrack, *best_sol.x))/star.log_R:.3f}',
+                                f'{100.0*abs(star.center_h1 - self.get_track_val("center_h1", htrack, *best_sol.x))/star.center_h1:.3f}',
+                                f'{100.0*abs(star.surface_h1 - self.get_track_val("surface_h1", htrack, *best_sol.x))/star.surface_h1:.4f}',
+                                f'{100.0*abs(star.he_core_mass - self.get_track_val("he_core_mass", htrack, *best_sol.x))/star.he_core_mass:.3f}',
+                                f'{100.0*abs(star.center_he4 - self.get_track_val("center_he4", htrack, *best_sol.x))/star.center_he4:.4f}',
+                                f'{100.0*abs(star.surface_he4 - self.get_track_val("surface_he4", htrack, *best_sol.x))/star.surface_he4:.4f}',    
+                                f'{100.0*abs(star.center_c12 - self.get_track_val("center_c12", htrack, *best_sol.x))/star.center_c12:.4f}'
```
minor: please check the 80 characters limit

mkruckow replied:
I'd remove the `abs` as the sign tell you the direction it is off. E.g. for the mass an increase (gain) is worth then a decrease (loss).

astroJeff commented:
@sgossage can you check to make sure the text in the v2 paper is consistent with the updates you have made here?

philipp-rajah reviewed (approved):
Reviewed for stylistic changes, looks good to me.

sgossage commented:
@astroJeff , I reviewed your changes and made some revisions to them to fix scope errors and also further functionalize things. Please have a look and see what you think.

maxbriel commented on posydon/binary_evol/binarystar.py:
```diff
@@ -364,10 +406,49 @@ def to_df(self, **kwargs):
+
+        def equalize_columns(data_to_save, all_keys, properties_dtypes,
+                             max_col_length):
+            """Iterate through data_to_save to append null values."""
```
```suggestion
            """Iterate through data_to_save to append null values.
            
            Parameters
            -----------
            data_to_save : list[str]
                Columns to save to the final dataframe
            all_keys : list[str]
                All available keys to store.
           properties_dtypes : dict
               Types of all available keys
           max_col_length: int
               The maximum length of columns.  
            """
```

sgossage commented:
> Nicely done! This looks really good and is nicely modular.
> 
> 1. I know this was in the code before, but in `__call__` of `step_detached`, there are 7 functions defined. Is it easy to move these out of the other function?
> 2. Same with `match_to_single_star` in `track_match.py`, where there are quite a few functions defined inside that one function.
>    For both, it would be great to separate the functions out.

Thanks, Max.

I've now cleaned out the `__call__` function in `step_detached` and isolated most of the functions. 

I've also cleaned up `match_to_single_star`, which is now its own function. Most of the functions that were nested there are now nested inside of the separate `match_through_minimization` function (b/c most of those subfunctions should only be used by the minimization routine); `match_through_root` is also its own function now.

sgossage commented:
> Thank you for the improvements! This looks good to me now. Eventually, we should move all functions outside of function definitions, because they're impossible to test. But for this PR it is fine to keep it as is. We will probably revisit this code in the future.
> 
> Is this a PR to be included in v2? Not yet right?

thanks, Max -- that's right, I believe we planned this for v2.1
